harvdev_utils/psycopg_functions/fb_feature_classes.py

```diff
@@ -498,10 +498,138 @@ class Insertion(Feature):
+        # Below are gene attributes that will be instantiated as "None" but given a value later.
+        self.fbal_list = None      # Will be a list of allele uniquenames.
+        self.floc_list = None      # Will be a list of (srcfeature_id, fmin, fmax, strand, FBrf ID) tuples.
+        self.fbrf_list = None      # Will be a list of FBrf pub IDs.
+        self.notes = None          # Will be a list of (string, FBrf ID) tuples.
+        # These attributes assessed by pick_floc() method after floc_list obtained.
+        self.chr_id = None         # Will be feature_id for Dmel chr/scaffold (golden_path(_region)).
+        self.fmin = None           # Will be featureloc.fmin (interbase).
+        self.fmax = None           # Will be featureloc.fmax (interbase).
+        self.strand = None         # Will be featureloc.strand.
+        self.agr_start = None      # Will be onbase "start" for export (adjusted as needed).
+        self.agr_stop = None       # Will be onbase "stop" for export (adjusted as needed).
+        # Attributes below require reference assembly info for determination.
+        self.chr_name = None       # Will be "uniquename" of the chromosome/scaffold.
+        self.chr_acc = None        # Will be the REFSEQ accession.version for the chromosome/scaffold.
+        self.ref_seq = None        # Will be the starting genomic sequence, usually "N/A" (string).
+        self.alt_seq = None        # Will be the variant genomic sequence, if known (string).
+        self.alt_acc = None        # Will be accession for inserted sequence, db:acc (string).
+        self.ins_len = None        # Will be the size of the insertion (integer).
+        self.padded_base = None    # Will be a string: A, C, G or T.
+        # Attributes below require synthesis of various Insertion data.
+        self.fails = []            # Will list various reasons object couldn't be exported.
+        self.agr_type = None       # Will be a Sequence Ontology term ID: e.g., usually "SO:0000667".
+        self.for_agr_export = None # Will be boolean.
 
     # feature.uniquename must be FBti-type.
     uniquename_regex = r'^FBti[0-9]{7}$'
 
+    def pick_floc(self):
+        """Determine featureloc specifics only if one featureloc entry found."""
+        if type(self.floc_list) is None:
+            log.warning('The "floc_list" is None - expected a list (even if empty).')
+        elif type(self.floc_list) != list:
+            log.warning('The "floc_list" is not a list - expected a list (even if empty).')
+        elif len(self.floc_list) == 0:
+            log.debug('Can\'t determine feature location for {}: no featureloc info in chado.'.format(self.uniquename))
+        elif len(self.floc_list) > 1:
+            log.debug('Can\'t determine feature location for {}: many featureloc values in chado.'.format(self.uniquename))
+        else:
+            self.chr_id = self.floc_list[0][0]
```

Comment:
Being picky I know. But can we have these 0-4 defined as constants and referenced here:-
self.chr_id = self.floc_list[0][CHR_ID] (or something) for all these.

Reply:
Yes, I see your point. I'm using a generic function that just adds sql query results (via psycopg2) to my objects as a list of tuples. I'll look into ways that I can apply labels for each tuple position, and turn it into a dict (key:value pairs). Does that seem ok?

Reply:
What I meant was to have something like 
CHR_ID = 0
and use that later which would be easier but there are other solutions.

Reply:
oh, so just within this method, specify CHR_ID = 0, FMIN = 1, etc first, then use those constants in the final "else" step. have I got it now?

Reply:
Yep, bingo you success.

Reply:
@ianlongden does this with some of the Python scrips and it makes it much readable when you look back at the code. I concur.

Reply:
For sure - already changed in my local branch